rfd/9999-webauthn-support.md

```diff
@@ -0,0 +1,501 @@
+AppID instead of the RP ID."_ This advice
+[wasn't followed by github.com/duo-labs/webauthn](https://github.com/duo-labs/webauthn/blob/9f1b88ef44cc0e4f5ddf511ed12a3aa468f972d7/protocol/assertion.go#L117),
+but in the short-term may be circumvented by creating a secondary
+[WebAuthn object](https://pkg.go.dev/github.com/duo-labs/webauthn/webauthn#WebAuthn)
+with RFID = AppID, used exclusively to validate legacy requests (identified by
+the presence of the "incorrect" rpIdHash).
+
+### UX and configuration
+
+WebAuthn support is designed to be as seamless as the U2F experience. Users
+enabling MFA for the first time may register their devices via `tsh mfa add` and
+use `tsh login` as usual. Users migrating from U2F to WebAuthn retain their
+previously-registered devices and keep their workflow unchanged.
+
+Migration from U2F to WebAuthn is achieved simply by changing the second_factor
+option from "u2f" to "webauthn". The WebAuthn backend sends challenges for all
+previously-registered U2F devices, in addition to any WebAuthn-registered
+devices, so migration is a seamless process - no re-registration required.
+
+If the second factor setting is "on", then the backend sends both U2F and
+WebAuthn challenges. In this case user-facing interfaces, such as `tsh` and the
+Proxy UI, should favor replying to WebAuthn challenges (instead of U2F).
+
+Additions to `teleport.yaml` are as follows:
+
+```yaml
+auth_service:
+  authentication:
+    type: local
+    second_factor: webauthn # "on" allows all MFA options
+    webauthn:
+      relying_party:
+        display_name: "Teleport Auth"
+        # ID of the Relying Party, ie, the Teleport service performing user
+        # authentication.
+        # Must be a domain name, such as "example.com".
+        # Defaults to the Proxy domain name.
+        # IMPORTANT: similarly to the U2F app ID, if the Relying Party ID
+        # changes then all devices must be re-registered.
+        id: localhost
+        # Origin of the Relying Party, validated against the origin of
+        # authentication requests.
+        # Defaults to the Proxy public address.
+        origin: https://localhost:3080
+        # URL which resolves to an image associated with the Relying Party.
+        # It must be an a priori authenticated URL of preferably 128 or less
+        # bytes, as authenticators may ignore an icon if its length is greater
+        # than 128 bytes.
+        icon: ""
+
+      # Previously configured U2F app ID, if applicable.
+      # Enables U2F-compatibility mode if set (explained below).
+      # The U2F app ID is sent as the `appid` WebAuthn extension for
+      # previously-registered U2F devices, to be used in place of the Relying
+      # Party ID, thus keeping backwards compatibility.
+      # Not required for new MFA installations.
+      # Defaults to "auth_service.authentication.u2f.app_id", if present.
+      u2f_app_id: "https://localhost:3080"
```

Comment:
I explicitly included this one because it allows users moving from u2f to webauthn to delete the entire u2f section of the config, keeping everything related to webauthn here.

Reply:
@codingllama what is the resource version of it? (we support two modes - static config and the dynamic resource for cloud and K8s and terraform). Is there a way for us to avoid extra field?

Reply:
The resource version for `AuthPreferenceSpecV2` doesn't have to change, we are only adding fields to it. Is that OK or do we have a different policy?  (The proto file leads me to believe we are at v2).

We could avoid the u2f_app_id field (more details about that in the "second_factor: u2f" thread). Taken to an extreme, we could avoid the entire webauthn configuration stanza if we are ok with keeping u2f.app_id indefinitely (and removing a few optional customization options, like the RP display_name and icon). As it stands it is optional anyway. I'd rather provide the toggles, but I won't lose any sleep if they're gone either  :)